### PR TITLE
raftstore: add hibernate-timeout to prevent leader become hibernated too fast 

### LINKED ISSUE #7747: Fail to evict leader when rolling update a new hibernated cluster
## Bug Report

If there are too many leaders or transfer leader just goes too slow, rolling update a cluster can take too much time. If the cluster enables hibernate region at the time, it's possible that all updated nodes will not have any leader because updated leader becomes hibernated very quickly and outdated follower will start election when lost heartbeat from leader for a long time. So it becomes slower because more leaders need to be evicted during rolling update.

A possible solution is to add a hibernate timeout configuration. Leader will not go to sleep before the timeout. So it's safe as long as the hibernate timeout is larger than the update time. And the timeout can be set to a smaller value after the cluster has enabled hibernate region on all nodes.

Another solution is make the configuration be changed online. So users can enable/disable hibernate regions whenever they feel safe.

We need to fix the issue in 3.0.x, 3.1.x and 4.0.

### PR DESCRIPTION
**Pull request by NingLin-P**
Signed-off-by: linning <linningde25@gmail.com>

### What problem does this PR solve?

Issue Number: close #7747 

Problem Summary:

Rolling update to hibernated cluster may take too long, because the hibernated leader will not send heartbeat to non-hibernated followers and cause them to start election.

### What is changed and how it works?

What's Changed:

Add `hibernate-timeout` config to raftstore, new elected leader will not be hibernated within this timeout, and this config can be changed online when the rolling update is finished.

### Release note 
- add hibernate-timeout config to prevent leader become hibernated too fast

### CODE REVIEW

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -980,8 +984,17 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
 
         self.fsm.peer.mut_store().flush_cache_metrics();
 
-        // Keep ticking if there are still pending read requests.
+        let new_elected_leader = match self.fsm.leader_elected_time.as_ref() {
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:987]: It may seem a little hacky, but how about calculating TiKV start time instead?

**Comment by NingLin-P** [comment]
Currently, we can not make the `hibernate-regions` configuration be changed online, because the online config change feature does not support changing all node's config at the same time like global config, and if some node enable `hibernate-regions` and other not, it just brings the same issue of rolling update, which we try to fix here.

Also, do you have any suggestion about the default value of `hibernate-timeout`? @BusyJay

File: components/raftstore/src/store/fsm/store.rs [hunk 2]
```diff
@@ -293,6 +294,20 @@ impl<T, C> PollContext<T, C> {
     pub fn store_id(&self) -> u64 {
         self.store.get_id()
     }
+
+    /// Timeout is calculated from TiKV start, the node should not become
+    /// hibernated if it still within the hibernate timeout, see
+    /// https://github.com/tikv/tikv/issues/7747
+    pub fn is_hibernate_timeout(&mut self) -> bool {
+        match self.node_start_time.as_ref() {
```
Comment (overvenus) [inline components/raftstore/src/store/fsm/store.rs:302]: What about just

```suggestion
        match self.node_start_time {
```

File: components/raftstore/src/store/fsm/store.rs [hunk 3]
```diff
@@ -293,6 +294,20 @@ impl<T, C> PollContext<T, C> {
...
+            Some(t) if t.elapsed() < self.cfg.hibernate_timeout.0 => false,
+            Some(_) => {
```
Comment (overvenus) [inline components/raftstore/src/store/fsm/store.rs:304]: Please avoid using `if` in match arm, sometimes it confuses people.

**Comment by NingLin-P** [comment]
PTAL @BusyJay

**Pull request merged**